docs/design/configs.md

```diff
@@ -0,0 +1,130 @@
+# Generic Config Object
+
+## Abstract
+
+A generic ``Config`` object to store command-line flags and environment variables
+that is used for configuration of ``Kubernetes`` components living on top of ``etcd``
+is proposed in this proposal.
```

Comment:
I still wonder whether the map[string] is right.  What about multi value strings?  Now you have to start slamming values together (so CLI supports -v a -v b, but the map is "a,b").  The simpler the pattern, the more people will abuse it (because it's convenient).  What about order of params?  Etc.  Basically, our config is complex because of "real world" - maybe we could reduce it in complexity, break them apart, etc, but then I think that generally applies to other types of config.  For secrets we chose to treat the values as opaque - do we get some value here in doing the same thing?

----- Original Message -----
> > @@ -0,0 +1,130 @@
> > +# Generic Config Object
> > +
> > +## Abstract
> > +
> > +A generic ``Config`` object to store command-line flags and environment
> > variables
> > +that is used for configuration of ``Kubernetes`` components living on top
> > of ``etcd``
> > +is proposed in this proposal.
> 
> That's a lot more complicated than I was envisioning when I first filed the
> issue.  I was thinking a map[string]string or maybe map[string]IntOrString
> - simply a substitute for flags.  Is the complexity really warranted?  We
> want to allow app users to use this too, right?
> 
> On Fri, Jun 19, 2015 at 2:01 PM, Clayton Coleman <notifications@github.com>
> wrote:
> 
> > In docs/design/configs.md
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6477#discussion_r32868986>
> > :
> >
> > > @@ -0,0 +1,130 @@
> > > +# Generic Config Object
> > > +
> > > +## Abstract
> > > +
> > > +A generic ``Config`` object to store command-line flags and environment
> > > variables
> > > +that is used for configuration of ``Kubernetes`` components living on
> > > top of ``etcd``
> > > +is proposed in this proposal.
> >
> > For reference, here's how OpenShift config objects are managed:
> > https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/v1/types.go
> > - they are versioned with their own scheme (like the kubectl cfg), have
> > their own Kinds, and support conversion and defaulting. We would prefer a
> > config solution that let these objects be managed equally well in etcd and
> > on disk. The configs also refer to secrets - today we use a secret with
> > many different keys to store the entire thing as a serialized directory for
> > running OpenShift on Kube. I think we would prefer stronger typing for
> > these objects where possible when we store them in etcd. We do have cases
> > where these configs are parameterized per instance (masters / nodes). As an
> > additional wrinkle, we allow direct component arguments to be passed
> > through -
> > https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/v1/types.go#L52
> > … <#14e0da054ab17164_>
> > ----- Original Message -----
> >  > @@ -0,0 +1,130 @@ > +# Generic Config Object > + > +## Abstract > + >
> > +A generic ``Config`` object to store command-line flags and environment >
> > variables > +that is used for configuration of ``Kubernetes`` components
> > living on top > of ``etcd`` > +is proposed in this proposal. * reduced
> > operational burden -- don't have to run your own etcd * simplified
> > deployment orchestration -- don't have to launch etcd, wait for it to
> > become ready, write config to etcd, then launch app * compatibility with
> > k8s tooling (e.g., kubectl create -f, labels, etc.) * use Kubernetes auth
> > and security policy * GC config when consumers are deleted -- if we
> > implement existence dependencies as proposed in #1535
> > <https://github.com/GoogleCloudPlatform/kubernetes/issues/1535> * consume
> > via file (if we support a volume, as with secrets) or env vars (if we
> > support that) rather than via API * for Kubernetes components, preserves
> > abstraction of the underlying storage mechanism (e.g., Kubelet doesn't need
> > to interact directly with etcd) --- Reply to this email directly or view it
> > on GitHub:
> > https://github.com/GoogleCloudPlatform/kubernetes/pull/6477/files#r32867440
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6477/files#r32868986>
> > .
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6477/files#r32870423